org.eclipse.xtext/src/org/eclipse/xtext/resource/containers/StateBasedContainer.java

```diff
@@ -110,7 +117,10 @@ public IResourceDescription getResourceDescription(URI uri) {
 	public Iterable<IEObjectDescription> getExportedObjectsByType(EClass type) {
 		if (isEmpty())
 			return emptyList();
-		return super.getExportedObjectsByType(type);
+
+		Predicate<IResourceDescription> isResourceDescritptionInState = getStateContainsPredicate(input -> input.getURI());
+		Iterable<IResourceDescription> resourceDescriptionsInState = Iterables.filter(getDescriptions().getAllResourceDescriptions(), isResourceDescritptionInState);
```

Comment:
this change wont use isResourceDescritptionInState (typo btw)

Reply:
its slightly better but still much worse than with the current version.
if we move

https://github.com/eclipse/xtext-core/blob/c2b20283fbd2057d9c4ae1b90ec907c813c7372e/org.eclipse.xtext/src/org/eclipse/xtext/resource/containers/StateBasedContainerManager.java#L87

to a factory method and expose
org.eclipse.xtext.resource.containers.StateBasedContainer.state
e.g. via a protected getter
then it could work.

even better woud be an addtional flag in the constructor that switches the behaviour.
but thats an ugly solution too.